### PR TITLE
more consistent marshal/unmarshal behavior for types that implement json Marshaler and Unmarshaler interfaces

### PR DESCRIPTION
**Pull request by jdef**
supersedes #302

### CODE REVIEW

File: jsonpb/jsonpb.go [hunk 1]
```diff
@@ -904,7 +904,7 @@ func (u *Unmarshaler) unmarshalValue(target reflect.Value, inputValue json.RawMe
 
 	// Handle nested messages.
 	if targetType.Kind() == reflect.Struct {
-		if target.CanAddr() {
+		if prop != nil && len(prop.CustomType) > 0 && target.CanAddr() { // jdef
```
Comment (awalterschulze) [inline jsonpb/jsonpb.go]: What is the comment for?
If you want you can add yourself to the CONTRIBUTORS or AUTHORS as appropriate.

**Comment by awalterschulze** [comment]
The rest looks perfect :)

File: jsonpb/jsonpb.go (hunk 1 above)
Comment (jdef) [inline jsonpb/jsonpb.go]: ugh, leftover bookmark. will remove

**Comment by awalterschulze** [comment]
Ok I am ready to merge this if you are?

**Comment by jdef** [comment]
:ship:

**Comment by awalterschulze** [comment]
Thank you for the fix :)

**Pull request merged**